### ISSUE TITLE
[2.7.x] Here is a type error when using ref() on attribute 'ref' in TSX

### ISSUE THREAD

**Issue by shiluo34** [vuejs/vue#12758]
### Version
2.7.10

### Reproduction link
[codesandbox.io](https://codesandbox.io/s/infallible-cloud-wnish2?file=/src/App.tsx)

### Steps to reproduce
Using type 'Ref' on ref will notice the error :  `jsx.d.ts(1305, 3): The expected type comes from property 'ref' which is declared here on type 'ElementAttrs<HTMLAttributes>'`. 

![image](https://user-images.githubusercontent.com/21286829/186367241-d1e80834-d06a-46a2-a11c-281b7e771dc6.png)

In the `vue/types/jsx.d.ts`, the type ReservedProps only support type'String' in Vue2.7.x，but supporting type 'String'/'Ref'/'Function' in Vue3.x

- Vue3 - jsx.d.ts
![image](https://user-images.githubusercontent.com/21286829/186367321-fe36171f-3f03-4104-b4cb-5a8c57e9a9d1.png "Vue3")

- Vue2.7.x -jsx.d.ts
![image](https://user-images.githubusercontent.com/21286829/186367413-e0e03b89-57c9-4fc0-836d-c512001c7bf7.png "Vue2")

### What is expected?
Using type 'Ref' on the property 'ref'.

### What is actually happening?
The type 'ref' only support 'String | undefined'.

**Comment by shiluo34** [comment]
fix pr: [#12759](https://github.com/vuejs/vue/pull/12759) :D

**Comment by onepisYa** [comment]
I also had the same problem

**Comment by shiluo34** [comment]
> I also had the same problem

yep, but this PR has`t been merged yet, i dont know whether something wrong with the PR or not :(

**Comment by onepisYa** [comment]
> > I also had the same problem
> 
> yep, but this PR has`t been merged yet, i dont know whether something wrong with the PR or not :(

Had the same problem using ts and [function-refs](https://vuejs.org/guide/essentials/template-refs.html#function-refs) in vue 2.7.x.

**Comment by shiluo34** [comment]
> 



> > > I also had the same problem
> > 
> > 
> > yep, but this PR has`t been merged yet, i dont know whether something wrong with the PR or not :(
> 
> Had the same problem using ts and [function-refs](https://vuejs.org/guide/essentials/template-refs.html#function-refs) in vue 2.7.x.
> 
> example:
> 
> `<input :ref="(el) => { /* assign el to a property or ref */ }">`

When I intend to modify this pr to add funciton ref type, I found  someone has already created the pr about this [#12763](https://github.com/vuejs/vue/pull/12763)

**Comment by onepisYa** [comment]
> > 
> 
> > > > I also had the same problem
> > > 
> > > 
> > > yep, but this PR has`t been merged yet, i dont know whether something wrong with the PR or not :(
> > 
> > 
> > Had the same problem using ts and [function-refs](https://vuejs.org/guide/essentials/template-refs.html#function-refs) in vue 2.7.x.
> > example:
> > `<input :ref="(el) => { /* assign el to a property or ref */ }">`
> 
> When I intend to modify this pr to add function ref type, I found someone has already created the pr about this [#12763](https://github.com/vuejs/vue/pull/12763)

Even after these two fixes, the el in the [function-ref](https://vuejs.org/guide/essentials/template-refs.html#function-refs) example still hints that implicitly has type any.

Writing functions directly in template can effectively use closures to access variables that are traversed by v-for. But it is not accessible in setup or methods options, so how should this problem be solved.

**Comment by shiluo34** [comment]
> > > 
> > 
> > 
> > > > > I also had the same problem
> > > > 
> > > > 
> > > > yep, but this PR has`t been merged yet, i dont know whether something wrong with the PR or not :(
> > > 
> > > 
> > > Had the same problem using ts and [function-refs](https://vuejs.org/guide/essentials/template-refs.html#function-refs) in vue 2.7.x.
> > > example:
> > > `<input :ref="(el) => { /* assign el to a property or ref */ }">`
> > 
> > 
> > When I intend to modify this pr to add function ref type, I found someone has already created the pr about this [#12763](https://github.com/vuejs/vue/pull/12763)
> 
> Even after these two fixes, the el in the [function-ref](https://vuejs.org/guide/essentials/template-refs.html#function-refs) example still hints that implicitly has type any.
> 
> Writing functions directly in template can effectively use closures to access variables that are traversed by v-for. But it is not accessible in setup or methods options, so how should this problem be solved.

I have updated my pr with fixing the function-refs type error. Defined type `Function` cant fix the question in #12763. I refer to the type define of  [vue3 ref type definition](https://github.com/vuejs/core/blob/fc5bdb36ed429d6c3c956f373206ce75467adaf3/packages/runtime-core/src/vnode.ts#L69). But you should define the el type when using the funtion refs, 
just like 
``` javascript
const fooRef = ref<HTMLElement>()

<div ref={(el) => {fooRef.value = el as HTMLElement}} /> 
```
its same as vue3. :D
